fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+                layer_handle = layer_type.register_full_backward_hook(helper.scale_gradients)
+                self._handles.append(layer_handle)
+                logging.debug("name = %s \t scale = %s" % (layer_name, elt.scaling_factor))
+
+    def unscale(self) -> None:
+        """
+        For each layer, check if any of the layers' parameters contain an inf/nan.
+        If there are no inf/nan in the gradient, then gradient of that layer is
+        unscaled by the reciprocal of the scaling factor for that layer.
+        Finally, all handles recorded while registering the hooks are deleted.
+        """
+        if self._apply_layerwise_scaling:
+            for elt in self.layer_info:
+                if not elt.found_inf_or_nan:
+                    for param_name, param in elt.layer_type.named_parameters():
+                        if hasattr(param, "grad"):
+                            logging.debug(
+                                "%s scaling down %s by %s" % (elt.layer_name, param_name, 1.0 / elt.scaling_factor)
+                            )
+                            param.grad = torch.mul(param.grad, 1.0 / elt.scaling_factor)
+
+            while len(self._handles) > 0:
+                elt = self._handles.pop()
+                elt.remove()
+
+    def _check_for_inf_or_nan(self) -> None:
+        """
+        For each layer, check if any of the parameters with a gradient attribute
+        contain an inf/nan. If any of the parameters' gradient contain an inf/nan,
+        then that layers' found_inf_or_nan attribute is set to True and all
+        remaining parameters for that layer are skipped.
+        """
+        for elt in self.layer_info:
+            elt.found_inf_or_nan = False
+            for _, param in elt.layer_type.named_parameters():
+                if hasattr(param, "grad") and param.grad is not None:
+                    if torch.isinf(param.grad).any().item() is True or torch.isnan(param.grad).any().item() is True:  # type: ignore
+                        elt.found_inf_or_nan = True
+                        break  # skip all remaining named parameters
+
+    def step(self, optimizer) -> None:  # type: ignore
+        """
+        If there are NO inf/nan in the gradients' of ALL layers, then optimizer
+        takes a step, otherwise not. Update the scaling factor for each layer.
+        """
+        # using layerwise gradient scaling
+        if self._apply_layerwise_scaling:
+            self._check_for_inf_or_nan()
+            inf_nan_found = any(elt.found_inf_or_nan for elt in self.layer_info)
+
+            if not inf_nan_found:
+                optimizer.step()
+            self._update_scale()
+        # not using layerwise gradient scaling
+        else:
+            optimizer.step()
+
+    def _update_scale(self) -> None:
+        """
+        For each layer, if an inf/nan is found, then multiply the scaling factor
```

Comment:
Can you explain what is the logic behind this? I think you are explaining the code below but what is useful is the reasoning behind how we are attempting to update the scale.

Reply:
If there is an inf/nan, then we should reduce the scale as the current scale is causing a gradient overflow. 

On the other hand if the gradient is finite we can increase the scale, but doing so at every step would again lead to an overflow within a short number of steps so we increase the scale gradually by pacing it. 
The growth interval determines how frequently we would increase the scale and the growth tracker measures where we are in relation to the growth interval.